airflow/ti_deps/ti_deps.py

```diff
@@ -0,0 +1,479 @@
+from abc import ABCMeta, abstractmethod
+from collections import namedtuple
+from datetime import datetime, timedelta
+from sqlalchemy import case, func
+
+import airflow
+from airflow.utils.db import provide_session
+from airflow.utils.state import State
+
+
+class BaseTIDep(object):
+    """
+    A dependency that must be satisfied in order for task instances to run. For example, a task can
+    only run if a certain number of its upstream tasks succeed.
+    """
+    __metaclass__ = ABCMeta
+
+    # TODO(aoen): when python 2.x is deprecated add the @abstractmethod decorator here (as
+    # combining abstractmethod and classmethod are done differently in 2.x and 3.x)
+    # TODO(aoen): All of the parameters other than the task instance should be replaced with an
+    # single context object parameter.
+    @classmethod
+    def get_dep_statuses(
+            cls,
+            ti,
+            session,
+            include_queued,
+            ignore_depends_on_past,
+            flag_upstream_failed):
+        """
+        Returns an iterable of TIDepStatus objects that describe whether the given task instance has
+        this dependency met.
+
+        For example a subclass could return an iterable of TIDepStatus objects, each one
+        representing if each of the passed in task's upstream tasks succeeded or not.
+
+        :param ti: the task instance to get the dependency status for
+        :type ti: TaskInstance
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def is_met(
+            cls,
+            ti,
+            session,
+            include_queued,
+            ignore_depends_on_past,
+            flag_upstream_failed):
+        """
+        Returns whether or not this dependency is met for a given task instance. A dependency is
+        considered met if all of the dependency statuses it reports are passing.
+
+        :param ti: the task instance to see if this dependency is met for
+        :type ti: TaskInstance
+        """
+        return all(status.passed for status in
+                   cls.get_dep_statuses(
+                       ti,
+                       session,
```

Comment:
You can look into the `@utils.provide_session` decorator that basically allows you to pass an active session, but if you don't is instantiate one and takes care of closing it at the end.